File: meshchatx/src/backend/active_sessions.py
# SPDX-License-Identifier: 0BSD

"""Track active UI WebSocket sessions (IP and user-agent) for local multi-client warnings."""

from __future__ import annotations

import ipaddress
import threading
import time
import uuid
from typing import Any

_MAX_UA_LEN = 512
_MAX_IP_LEN = 128


def _clean_ip(value: str | None) -> str:
    cleaned = str(value or "").strip()
    if not cleaned:
        return "unknown"
    return cleaned[:_MAX_IP_LEN]


def _clean_user_agent(value: str | None) -> str:
    cleaned = str(value or "").strip()
    if not cleaned:
        return "unknown"
    # Strip control characters that break logs or JSON displays.
    cleaned = "".join(ch for ch in cleaned if ch.isprintable())
    if not cleaned:
        return "unknown"
    return cleaned[:_MAX_UA_LEN]


def is_loopback_or_lan_ip(value: str | None) -> bool:
    """True for loopback, RFC1918, link-local, and IPv6 ULA addresses."""
    cleaned = str(value or "").strip()
    if not cleaned or cleaned.lower() == "unknown":
        return False
    if cleaned.startswith("[") and cleaned.endswith("]"):
        cleaned = cleaned[1:-1]
    if "%" in cleaned:
        cleaned = cleaned.split("%", 1)[0]
    if cleaned.lower().startswith("::ffff:"):
        cleaned = cleaned[7:]
    try:
        addr = ipaddress.ip_address(cleaned)
    except ValueError:
        return False
    return bool(addr.is_loopback or addr.is_private or addr.is_link_local)


def sessions_are_local_only(sessions: Any) -> bool:
    """True when every session IP is loopback or LAN (or the list is empty)."""
    if sessions is None:
        return False
    try:
        rows = list(sessions)
    except TypeError:
        return False
    if not rows:
        return True
    for row in rows:
        if not isinstance(row, dict):
            return False
        if not is_loopback_or_lan_ip(row.get("ip")):
            return False
    return True


def should_warn_multi_session(
    count: int,
    warning_enabled: bool,
    sessions: Any = None,
) -> bool:
    """Oracle: warn when two or more non-local sessions are active and the setting is on.

    Localhost and LAN-only clients (same machine or private network) do not
    trigger the toast. Pass sessions=None to skip the locality check.
    """
    try:
        active = int(count)
    except (TypeError, ValueError):
        return False
    if not bool(warning_enabled) or active < 2:
        return False
    if sessions is not None and sessions_are_local_only(sessions):
        return False
    return True


class ActiveSessionTracker:
    """In-memory registry of connected MeshChatX UI WebSocket clients."""

    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._sessions: dict[str, dict[str, Any]] = {}

    def add(self, *, ip: str | None, user_agent: str | None) -> dict[str, Any]:
        session_id = uuid.uuid4().hex
        entry = {
            "id": session_id,
            "ip": _clean_ip(ip),
            "user_agent": _clean_user_agent(user_agent),
            "connected_at": time.time(),
        }
        with self._lock:
            self._sessions[session_id] = entry
            return dict(entry)

    def remove(self, session_id: str | None) -> bool:
        cleaned = str(session_id or "").strip()
        if not cleaned:
            return False
        with self._lock:
            return self._sessions.pop(cleaned, None) is not None

    def count(self) -> int:
        with self._lock:
            return len(self._sessions)

    def list_sessions(self) -> list[dict[str, Any]]:
        with self._lock:
            rows = [dict(row) for row in self._sessions.values()]
        rows.sort(key=lambda row: float(row.get("connected_at") or 0.0))
        return rows

    def snapshot(self) -> dict[str, Any]:
        sessions = self.list_sessions()
        return {
            "count": len(sessions),
            "sessions": sessions,
        }

File: meshchatx/src/frontend/js/activeSessions.ts
// SPDX-License-Identifier: 0BSD

/**
 * Predict whether the multi-session warning toast should fire.
 * Mirrors meshchatx.src.backend.active_sessions.should_warn_multi_session.
 */

function parseIpv4Parts(value) {
    const match = String(value || "").match(/^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/);
    if (!match) {
        return null;
    }
    const parts = match.slice(1).map((part) => Number(part));
    if (parts.some((part) => !Number.isInteger(part) || part < 0 || part > 255)) {
        return null;
    }
    return parts;
}

/**
 * True for loopback, RFC1918, link-local, and IPv6 ULA addresses.
 * Mirrors meshchatx.src.backend.active_sessions.is_loopback_or_lan_ip.
 */
export function isLoopbackOrLanIp(value) {
    let cleaned = String(value || "")
        .trim()
        .toLowerCase();
    if (!cleaned || cleaned === "unknown") {
        return false;
    }
    if (cleaned.startsWith("[") && cleaned.endsWith("]")) {
        cleaned = cleaned.slice(1, -1);
    }
    if (cleaned.includes("%")) {
        cleaned = cleaned.split("%")[0];
    }
    if (cleaned.startsWith("::ffff:")) {
        cleaned = cleaned.slice(7);
    }
    if (cleaned === "::1" || cleaned === "0:0:0:0:0:0:0:1") {
        return true;
    }

    const ipv4 = parseIpv4Parts(cleaned);
    if (ipv4) {
        const [a, b] = ipv4;
        if (a === 127) {
            return true;
        }
        if (a === 10) {
            return true;
        }
        if (a === 192 && b === 168) {
            return true;
        }
        if (a === 172 && b >= 16 && b <= 31) {
            return true;
        }
        if (a === 169 && b === 254) {
            return true;
        }
        return false;
    }

    if (cleaned.startsWith("fe80:")) {
        return true;
    }
    const firstHextet = cleaned.split(":", 1)[0] || "";
    if (/^f[cd][0-9a-f]{0,2}$/.test(firstHextet)) {
        return true;
    }
    return false;
}

/**
 * True when every session IP is loopback or LAN (or the list is empty).
 */
export function sessionsAreLocalOnly(sessions) {
    if (!Array.isArray(sessions)) {
        return false;
    }
    if (sessions.length === 0) {
        return true;
    }
    return sessions.every((row) => isLoopbackOrLanIp(row?.ip));
}

export function shouldWarnMultiSession(count, warningEnabled, sessions) {
    const active = Number(count);
    if (!Number.isFinite(active)) {
        return false;
    }
    if (warningEnabled === false || active < 2) {
        return false;
    }
    if (sessions !== undefined && sessionsAreLocalOnly(sessions)) {
        return false;
    }
    return true;
}

/**
 * Fire at most once per multi-session episode (count rises to 2+).
 * Resets when the count drops below 2 so a later episode can warn again.
 */
export function shouldShowMultiSessionToast(count, warningEnabled, alreadyWarned, sessions) {
    const shouldWarn = shouldWarnMultiSession(count, warningEnabled, sessions);
    if (!shouldWarn) {
        return { show: false, warned: false };
    }
    if (alreadyWarned) {
        return { show: false, warned: true };
    }
    return { show: true, warned: true };
}
